### PR TITLE
ARM64 - Always morph GT_MOD

### PR DESCRIPTION
**Pull request by TIHan**
Should resolve these regressions: 

- https://github.com/dotnet/runtime/issues/67983 - more info about it here: https://github.com/dotnet/runtime/pull/66407#issuecomment-1098370889
- ~~https://github.com/dotnet/runtime/issues/68470~~ Seems to not reproduce in main anymore.
- https://github.com/dotnet/runtime/issues/68136

**Description**

We only want to do the specific ARM64 mod optimization if the morphed `Mod` to `SubMulDiv` did not take advantage of CSE. The only phase, that I know of, to do any kind of transformation that occurs after CSE is 'lowering'.

The idea here is to find the `SubMulDiv` in lowering, turn it back into a `Mod`, and then call `LowerModPow2` to do the specific optimization.

But it gets more complicated. `SubMulDiv` itself gets optimized in lowering as well, so the shape of `SubMulDiv` is lost. Therefore, we need to find other possible shapes that get lowered from `SubMulDiv`.

**Acceptance Criteria**

- [x] Add Fuzzlyn regression tests
- [x] Merge https://github.com/dotnet/runtime/pull/69770

### CODE REVIEW

**Comment by TIHan** [comment]
@kunalspathak - I think this is handling those regressions we saw earlier with the use of `csnegs`. 

Looking at the asmdiffs now, it does show a lot of regressions, but they are caused by the additional divbyzero and overflow check:
```diff
G_M60555_IG04:        ; gcVars=0000000000000000 {}, gcrefRegs=0000 {}, byrefRegs=0000 {}, gcvars, byref
            bl      CORINFO_HELP_OVERFLOW
						;; size=4 bbWeight=0    PerfScore 0.00
G_M60555_IG05:        ; gcrefRegs=0000 {}, byrefRegs=0000 {}, byref
            bl      CORINFO_HELP_THROWDIVZERO
            brk_windows #0
						;; size=8 bbWeight=0    PerfScore 0.00
```
These were here before we merged `csnegs` so it's just putting them back. I would say these are not regressions.

**Comment by jakobbotsch** [comment]
I think recognizing patterns this large in lowering is too fragile and hard to get right.
How do we end up with these patterns? E.g. I see `(x >> k) << k` -- why is it not `x & ~((1 << k) - 1)` at this point, and if there is a good reason, can we add this transformation separately instead? Similarly to the other patterns we see here, is it possible to optimize them on their own instead, with the idea that hopefully we end up with something simpler to recognize/optimized on its own?

**Comment by jakobbotsch** [comment]
It might also be reasonable to introduce additional transformations as part of rationalization (or separately, with higher TP cost). We also have #68103 where it would be good to have. Thoughts @dotnet/jit-contrib?

**Comment by BruceForstall** [comment]
> I think recognizing patterns this large in lowering is too fragile and hard to get right.

I definitely agree with this. I don't know if there is a better option (why can't it be done in morph?), though.

**Comment by TIHan** [comment]
> I think recognizing patterns this large in lowering is too fragile and hard to get right.

I agree as well. Originally, I only wanted to find the `SubMulDiv` pattern. But at that point in lowering, that is what `SubMulDiv` turns into.

This specific optimization shouldn't be done in morph, even if we had the constructs to do so. The reason why is because the transformation of `a % b` to `a - (a / b) * b` - we want to take advantage of `a / b` for CSE. There is code out there with this pattern, which is used when you want to mark bits:
```fsharp
let x = index / 16
let y = index % 16
```
which would turn into:
```fsharp
let x = index / 16
let y = index - (index / 16) * 16
```
then:
```fsharp
let cse = index / 16
let y = index - cse * 16
```

What I really want to do is look for `SubMulDiv` after CSE is done and before `SubMulDiv` gets lowered to the much larger pattern. Where is the best place to do that? Do we need to introduce a new phase to achieve that?

**Comment by jakobbotsch** [comment]
> Where is the best place to do that? Do we need to introduce a new phase to achieve that?

I would be interested to see a prototype of the suggestions I had above. I.e. try it as part of rationalization (in pre-order there), and try a new late phase (e.g. after range check). We can see how costly the full IR walk will be and evaluate whether having the optimization and future opportunity to do other things in that pass outweighs that cost.

**Comment by SingleAccretion** [comment]
Note that we already have an "early lowering" pass in our phase order that is a (mostly "empty") full IR walk.

**Comment by jakobbotsch** [comment]
> Note that we already have an "simple lowering" pass in our phase order that is a (mostly "empty") full IR walk.

It's after rationalization however, so folding transformations still require interference checking here (or otherwise coupling it to rationalization).

**Comment by SingleAccretion** [comment]
Yep, the idea would be to move it to before rationalization. Tree walks are still costlier than linear traversals, but not "a new full IR walk" costlier at least.

File: src/coreclr/jit/rationalize.cpp
Comment (jakobbotsch) [inline src/coreclr/jit/rationalize.cpp:341]: So we are looking for
```
SUB      (named node)
    LCL_VAR (named op1)
    LSH            (named lsh)
        DIV        (named div)
            LCL_VAR (named a)
            CNS_INT (named cns2)
        CNS_INT (named cns1)
```
The evaluation order here depends on `GTF_REVERSE_OPS`, but just removing all the nodes you don't bash and leaving `cns2` in place should work correctly. In other words, I think this code should be

```cpp
node->ChangeOper(GT_MOD);
node->AsOp()->gtOp2 = cns2;
BlockRange().Remove(lsh, div, a, cns1);
```

File: src/coreclr/jit/rationalize.cpp
Comment (TIHan) [reply src/coreclr/jit/rationalize.cpp:633]: If it needs to be done in preorder, then there needs to be additional logic to handle the interference checks, right? Because putting it in pre-order won't match the desired shape in scenarios that have COMMAS.

Comment (jakobbotsch) [reply src/coreclr/jit/rationalize.cpp:633]: If it's done in post-order then it needs interference checks, precisely because you won't see the commas as operands -- those commas contain embedded statements that are linearized and that you need to check interference with in post-order.

In what situations are you seeing commas where the transform would be able to kick in? That sounds a little surprising to me.

Comment (TIHan) [reply src/coreclr/jit/rationalize.cpp:633]: [asmdiffs.zip](https://github.com/dotnet/runtime/files/8721040/asmdiffs.zip)

Comment (jakobbotsch) [reply src/coreclr/jit/rationalize.cpp:633]: FWIW, it's not hard to do the interference check that I expect will catch all cases we care about -- you can just check if the range of the `sub` is closed.

Comment (TIHan) [reply src/coreclr/jit/rationalize.cpp:633]: Example program:
```csharp
using System;
using System.Runtime.CompilerServices;

namespace Benchstone.BenchI
{
    public static class TreeSort
    {

#if DEBUG
    public const int Iterations = 1;
#else
        public const int Iterations = 1200;
#endif

        const int SortElements = 5000;
        const int Modulus = 65536;

        sealed class Node
        {
            public Node Left;
            public Node Right;
            public int Val;
            public Node(int n)
            {
                Left = null;
                Right = null;
                Val = n;
            }
        }

        static int s_biggest;
        static int s_littlest;
        static int s_seed;

        static void InitRand()
        {
            s_seed = 33;
        }

        static int Rand(ref int seed)
        {
            int multiplier = 25173;
            int increment = 13849;
            seed = (multiplier * seed + increment) % Modulus;
            return seed;
        }

        static void InitArray(int[] sortList)
        {
            InitRand();
            s_biggest = 0;
            s_littlest = Modulus;
            for (int i = 1; i <= SortElements; i++)
            {
                sortList[i] = Rand(ref s_seed) - 1;
                if (sortList[i] > s_biggest)
                {
                    s_biggest = sortList[i];
                }
                else if (sortList[i] < s_littlest)
                {
                    s_littlest = sortList[i];
                }
            }
        }

        static void Insert(int n, Node t)
        {
            if (n > t.Val)
            {
                if (t.Left == null)
                {
                    t.Left = new Node(n);
                }
                else
                {
                    Insert(n, t.Left);
                }
            }
            else if (n < t.Val)
            {
                if (t.Right == null)
                {
                    t.Right = new Node(n);
                }
                else
                {
                    Insert(n, t.Right);
                }
            }
        }

        static bool CheckTree(Node p)
        {
            bool result = true;
            if (p.Left != null)
            {
                if (p.Left.Val <= p.Val)
                {
                    result = false;
                }
                else
                {
                    result &= CheckTree(p.Left);
                }
            }

            if (p.Right != null)
            {
                if (p.Right.Val >= p.Val)
                {
                    result = false;
                }
                else
                {
                    result &= CheckTree(p.Right);
                }
            }

            return result;
        }

        static bool Trees(int[] sortList)
        {
            InitArray(sortList);
            Node tree = new Node(sortList[1]);
            for (int i = 2; i <= SortElements; i++)
            {
                Insert(sortList[i], tree);
            }
            bool result = CheckTree(tree);
            return result;
        }

        [MethodImpl(MethodImplOptions.NoInlining)]
        static bool Bench()
        {
            int[] sortList = new int[SortElements + 1];
            bool result = Trees(sortList);
            return result;
        }

        static bool TestBase()
        {
            bool result = true;
            for (int i = 0; i < Iterations; i++)
            {
                result &= Bench();
            }
            return result;
        }

        public static int Main()
        {
            bool result = TestBase();
            return (result ? 100 : -1);
        }
    }
}

```

Looking at part of `InitArray` :
```diff
 G_M14667_IG04:        ; gcrefRegs=0000 {}, byrefRegs=0028 {x3 x5}, byref
             str     w0, [x1,#-16]
             b       G_M14667_IG06
@@ -104,12 +104,15 @@ G_M14667_IG08:        ; gcrefRegs=0001 {x0}, byrefRegs=0000 {}, byref, isz
             mov     w5, #0xd1ffab1e
             mov     w6, #0xd1ffab1e
             madd    w4, w4, w5, w6
-            and     w5, w4, #0xd1ffab1e
-            negs    w4, w4
-            and     w4, w4, #0xd1ffab1e
-            csneg   w4, w5, w4, mi
+            asr     w5, w4, #31
+            and     w5, w5, #0xd1ffab1e
+            add     w5, w5, w4
+            asr     w5, w5, #16
+            lsl     w5, w5, #16
+            sub     w4, w4, w5
             str     w4, [x3]
-            sub     w3, w4, #1
+            ldr     w3, [x3]
+            sub     w3, w3, #1
             mov     w4, w3
             ldr     w5, [x0,#8]
             cmp     w2, w5
@@ -121,7 +124,7 @@ G_M14667_IG08:        ; gcrefRegs=0001 {x0}, byrefRegs=0000 {}, byref, isz
             ; byrRegs -[x5]
             cmp     w3, w5
             ble     G_M14667_IG10
-						;; size=88 bbWeight=0.04 PerfScore 0.88
+						;; size=100 bbWeight=0.04 PerfScore 1.10
```

Comment (jakobbotsch) [reply src/coreclr/jit/rationalize.cpp:633]: BTW, you can do this in `fgSimpleLowering` instead without having to move that phase. That makes the change a little less controversial :)

Comment (TIHan) [reply src/coreclr/jit/rationalize.cpp:573]: Doing it in simple lowering makes sense, though the optimization isn't exactly "simple" :)

**Comment by TIHan** [comment]
@dotnet/jit-contrib This is ready. Will post diffs when CI finishes.

There are just a few regressions, mostly due to extra CSEs so they should not actually be regressions.

However, I did find one interesting regression for `System.Text.RegularExpressions.Symbolic.BitVector:CreateTrue(int):System.Text.RegularExpressions.Symbolic.BitVector`:
```diff
@@ -7,7 +7,7 @@
 ; 0 inlinees with PGO data; 1 single block inlinees; 1 inlinees without PGO data
 ; Final local variable assignments
 ;
-;  V00 RetBuf       [V00,T04] (  3,  3   )   byref  ->  x19         single-def
+;  V00 RetBuf       [V00,T03] (  3,  3   )   byref  ->  x19         single-def
 ;  V01 arg0         [V01,T01] (  4,  4   )     int  ->   x0         single-def
 ;  V02 loc0         [V02,T00] (  7,  6.50)  struct (24) [fp+18H]   do-not-enreg[SF] must-init ld-addr-op
 ;# V03 OutArgs      [V03    ] (  1,  1   )  lclBlk ( 0) [sp+00H]   "OutgoingArgSpace"
@@ -17,8 +17,8 @@
 ;  V07 tmp4         [V07,T06] (  3,  3   )     ref  ->  x13         single-def "arr expr"
 ;  V08 tmp5         [V08,T07] (  3,  3   )     int  ->  x20         "index expr"
 ;  V09 cse0         [V09,T09] (  3,  2.50)     int  ->  x20         "CSE - aggressive"
-;  V10 rat0         [V10,T02] (  3,  6   )     int  ->   x1         "ReplaceWithLclVar is creating a new local variable"
-;  V11 rat1         [V11,T03] (  3,  6   )     int  ->  x14         "ReplaceWithLclVar is creating a new local variable"
+;  V10 cse1         [V10,T04] (  4,  4   )     int  ->  x14         "CSE - aggressive"
+;  V11 rat0         [V11,T02] (  3,  6   )     int  ->   x1         "ReplaceWithLclVar is creating a new local variable"
 ;
 ; Lcl frame size = 32
 
@@ -55,12 +55,14 @@ G_M28410_IG02:        ; gcrefRegs=0000 {}, byrefRegs=80000 {x19}, byref, isz
             blr     x2
             ; gcrRegs -[x0]
             ldr     w14, [fp,#32]
-            and     w13, w14, #63
-            negs    w14, w14
-            and     w14, w14, #63
-            csneg   w14, w13, w14, mi
+            asr     w13, w14, #31
+            and     w13, w13, #63
+            add     w13, w13, w14
+            asr     w13, w13, #6
+            lsl     w13, w13, #6
+            sub     w14, w14, w13
             cbz     w14, G_M28410_IG04
-						;; size=108 bbWeight=1    PerfScore 23.00
+						;; size=116 bbWeight=1    PerfScore 25.50
```

My guess is that a `GT_COMMA` was introduced (probably the CSE) as one of the operands, which breaks matching the pattern in Preorder rationalization. But, this technically isn't a regression as the code is not worse than before we merged the mod optimization using `csneg`.

File: src/tests/JIT/Regression/JitBlue/GitHub_68136/GitHub_68136.cs
Comment (jakobbotsch) [inline src/tests/JIT/Regression/JitBlue/GitHub_68136/GitHub_68136.cs:1]: This needs a license header.

Comment (jakobbotsch) [reply src/tests/JIT/Regression/JitBlue/GitHub_68136/GitHub_68136.cs:1]: Also, you should rename the file and class name here to `Runtime_68136` (the GitHub_XYZ issues refer to issues when the repo was at dotnet/coreclr).

File: src/coreclr/jit/rationalize.cpp
Comment (jakobbotsch) [inline src/coreclr/jit/rationalize.cpp:289]:
```suggestion
    if (!(node->TypeIs(TYP_INT, TYP_LONG) && op1->OperIs(GT_LCL_VAR)))
```

This is always true, `GT_SUB` cannot be unsigned.

Comment (jakobbotsch) [reply src/coreclr/jit/rationalize.cpp:289]: Good catch, yeah it should rather be changed to `!node->gtOverflow()`.

Comment (TIHan) [reply src/coreclr/jit/rationalize.cpp:289]: Does overflow matter here? When this is fully transformed, the divisor has to be a constant power of 2 which could never overflow.

Comment (TIHan) [reply src/coreclr/jit/rationalize.cpp:289]: Does overflow matter here? When this is fully transformed, the divisor has to be a constant power of 2 which could never overflow.

Comment (jakobbotsch) [reply src/coreclr/jit/rationalize.cpp:289]: No, this probably cannot overflow. But in that case I still think the `IsUnsigned()` check does not make much sense.

File: src/coreclr/jit/rationalize.cpp
Comment (jakobbotsch) [inline src/coreclr/jit/rationalize.cpp:312]: I don't think this is right if `lsh` has `GTF_REVERSE_OPS` set. I think to handle `GTF_REVERSE_OPS` properly the easiest will be to remove all the nodes individually.

It would be very helpful to add a comment showing the full tree that is being changed here.

File: src/coreclr/jit/rationalize.cpp
Comment (jakobbotsch) [reply src/coreclr/jit/rationalize.cpp:313]: Ah, I suppose it gets removed by the `Remove` call above.

File: src/coreclr/jit/rationalize.cpp
Comment (TIHan) [reply src/coreclr/jit/rationalize.cpp:274]: Yes, it should be

File: src/coreclr/jit/rationalize.cpp
Comment (TIHan) [reply src/coreclr/jit/rationalize.cpp:312]: I'm basically going to bail if I see any of the ops having the reverse op set.

For adding a comment, is `a - (a / cns) << shift  =>  a % cns` not sufficient? I could add the full tree here, but it feels redundant.

Comment (jakobbotsch) [reply src/coreclr/jit/rationalize.cpp:319]: > For adding a comment, is a - (a / cns) << shift => a % cns not sufficient? I could add the full tree here, but it feels redundant.

I'll leave it up to you, I personally would prefer the detailed tree form but if you prefer the simplified form then that's totally fine with me.

I still do think the code here is unnecessarily complicated. Why not just remove only the nodes that you don't reuse? That way there is no need to resequence the tree or find the first node, plus no need to be conservative about `GTF_REVERSE_OPS` because removing the nodes individually handles that too. I.e. the entirety of the block becomes:
```cpp
node->ChangeOper(GT_MOD);
node->AsOp()->gtOp2 = cns;
BlockRange().Remove(lsh);
BlockRange().Remove(div);
BlockRange().Remove(a);
BlockRange().Remove(shift);
```
if I'm not mistaken.

BTW, should we guard this transform under `comp->opts.OptimizationEnabled()`?

Comment (TIHan) [reply src/coreclr/jit/rationalize.cpp:319]: > I still do think the code here is unnecessarily complicated.

My head still has a hard time when it comes to lowering when I have to worry about the trees and the linear form at the same time; that in of itself is already complicated. So, I'm not surprised if what I have is complicated :) - you notice I keep going back and forth between different versions of trying to do the lowering. This is the reason why I was experimenting with a new phase right before rationalization where I would only have to worry about the trees; less to worry/think about.

> comp->opts.OptimizationEnabled()

Wouldn't hurt to do so.

File: src/coreclr/jit/rationalize.cpp
Comment (TIHan) [reply src/coreclr/jit/rationalize.cpp:289]: If `GT_SUB` is guaranteed to never be `unsigned` then I guess we can remove the check.

File: src/coreclr/jit/lowerarmarch.cpp
Comment (TIHan) [inline src/coreclr/jit/lowerarmarch.cpp:107]: Needed to add this so that `GT_CMP`'s second operand can be properly contained.

**Comment by TIHan** [comment]
@dotnet/jit-contrib this is ready again, CI is green now

**Comment by TIHan** [comment]
[Diffs](https://dev.azure.com/dnceng/public/_build/results?buildId=1816320&view=ms.vss-build-web.run-extensions-tab)

**Comment by jakobbotsch** [comment]
I'd suggest to split the new optimization into a separate PR so we can review it and track its impact separately.

File: src/coreclr/jit/rationalize.cpp
Comment (jakobbotsch) [inline src/coreclr/jit/rationalize.cpp:299]: Nit: I don't think we need to bail in this case, the logic below should handle it correctly.

Comment (TIHan) [reply src/coreclr/jit/rationalize.cpp]: done

**Comment by TIHan** [comment]
@jakobbotsch I split them out

File: src/coreclr/jit/lowerarmarch.cpp
Comment (jakobbotsch) [inline src/coreclr/jit/lowerarmarch.cpp:730]: Are these changes necessary or should they be done in the new PR too?

File: src/coreclr/jit/morph.cpp
Comment (jakobbotsch) [inline src/coreclr/jit/morph.cpp:10577]: Ditto, should be done in the new PR?

File: src/coreclr/jit/lowerarmarch.cpp
Comment (TIHan) [reply src/coreclr/jit/lowerarmarch.cpp:730]: Sort of? I got rid of the `use`. New PR extends this path anyway with the new instruction.

File: src/coreclr/jit/morph.cpp
Comment (TIHan) [reply src/coreclr/jit/morph.cpp:10577]: It should be done in this PR.

File: src/coreclr/jit/lowerarmarch.cpp
Comment (jakobbotsch) [reply src/coreclr/jit/lowerarmarch.cpp:730]: But this is unrelated to the fix, right?
I am fine with leaving it to avoid more churn on this PR.

**Review by jakobbotsch — APPROVED** [review]

File: src/coreclr/jit/lowerarmarch.cpp
Comment (TIHan) [reply src/coreclr/jit/lowerarmarch.cpp:730]: It is unrelated, been trying different ways to shape this function. The next PR is the shape  and impl I'm happy with.

**Pull request merged**
